edi_shipment.py

```diff
@@ -0,0 +1,615 @@
+                [('suppliers.type_', '=', 'NADSU'),
+                    ('suppliers.party', ) + tuple(clause[1:])]]
+
+    def get_party(self, name):
+        if self.manual_party:
+            return self.manual_party.id
+        for s in self.suppliers:
+            if s.type_ == 'NADSU':
+                return s.party and s.party.id
+
+    def read_BGM(self, message):
+        self.number = message.pop(0)
+        self.type_ = message.pop(0)
+        self.function_ = message.pop(0)
+
+    def read_DTM(self, message):
+        if message:
+            self.expedition_date = to_date(message.pop(0))
+        if message:
+            self.estimated_date = to_date(message.pop(0))
+
+    def read_RFF(self, message):
+        REF = Pool().get('edi.shipment.in.reference')
+        ref = REF()
+        ref.type_ = message.pop(0)
+        if message:
+            ref.reference = message.pop(0)
+        if message:
+            ref.reference_date = to_date(message.pop(0))
+        ref.search_reference()
+        if not getattr(self, 'references', False):
+            self.references = []
+        self.references += (ref,)
+
+    def read_TOD(self, message):
+        # Not implemented
+        pass
+
+    def read_TDT(self, message):
+        # Not implemented
+        pass
+
+    def read_CPS(self, message):
+        # Not implemented
+        pass
+
+    def read_PAC(self, message):
+        # Not implemented
+        pass
+
+    def read_HAN(self, message):
+        # Not implemented
+        pass
+
+    def read_PCI(self, message):
+        # Not implemented
+        pass
+
+    def read_ALI(self, message):
+        # Not implemented
+        pass
+
+    def read_CNTRES(self, message):
+        # Not implemented
+        pass
+
+    def read_MOA(self, message):
+        # Not implemented
+        pass
+
+    def read_MEA(self, message):
+        # Not implemented
+        pass
+
+    def get_quantity(line):
+        for qty in line.quantities:
+           if qty.type_ == '12':
+               return float(qty.quantity)
+
+
+    @classmethod
+    def import_edi_file(cls, shipments, data):
+        pool = Pool()
+        ShipmentEdi = pool.get('edi.shipment.in')
+        ShipmentEdiLine = pool.get('edi.shipment.in.line')
+        SupplierEdi = pool.get('edi.shipment.supplier')
+        # Configuration = pool.get('stock.configuration')
+
+        # config = Configuration(1)
+        separator = '|'  # TODO config.separator
+
+        shipment_edi = None
+        shipment_line = None
+        document_type = data.pop(0).replace('\n', '').replace('\r', '')
+        if document_type != 'DESADV_D_96A_UN_EAN005':
+            return
+        for line in data:
+            line = line.replace('\n', '').replace('\r', '')
+            line = line.split(separator)
+            msg_id = line.pop(0)
+            if msg_id == 'BGM':
+                shipment_edi = ShipmentEdi()
+                shipment_edi.read_BGM(line)
```

Comment:
Afegiria aquí:

    shipment_edi.lines = []

i ens estalviem més avall fer un getattr:

    if not getattr(shipment_edi, 'lines', False):

Idem pel:

  shipment_edi.suppliers = []

També quan es crea aquest objecte, trobo a faltar el company. Si fem un default_values, podem obtenir ja els valors per defecte d'aquest objecte.

    shipment_edi_line = ShipmentEdiLine(**default_values)